man/man1/txt2clm.1
.TH txt2clm 1  "USER COMMANDS"
.SH NAME
txt2clm \- convert text files to clm data files for LPJmL
.SH SYNOPSIS
.B txt2clm
[\-h] [\-version v] [\-float] [\-int] [\-csv c] [\-scale s] [\-nbands n] [-nstep n] [\-timestep t] [\-firstyear f] [\-firstcell c] [\-ncell n]  [\-header id] [\-json]
.I txtfile clmfile
.SH DESCRIPTION
Program converts text files containing white space separated numbers into LPJ climate data files. File is read until end of file is reached.
.SH OPTIONS
.TP
\-h,\--help
Print a short help text.
.TP
\-version v
Set clm version, default is 3.
.TP
\-float
Write data as float, default is short.
.TP
\-int
Write data as int, default is short.
.TP
\-csv c
Text file is in CSV format with separator character c.
.TP
\-scale s
Scale data by a factor.
.TP
\-nbands n
Set number of bands, default is 12.
.TP
\-nstep n
Set number of time steps within year, default is 1. Number of bands and number of steps can both by >1 only if version of clm file is set to 4.
TP
\-timestep t
Set number of years between time steps, default is 1. If timestep is greater than 1 then number of time steps within a year must be 1.
.TP
\-firstyear f
Set first year, default is 1901.
.TP
\-firstcell c
Set first cell, default is 0.
.TP
\-ncell n
Set number of cells, default is 67420.
.TP
\-header id
Set header string in clm file, default is LPJCLIM.
.TP
\-json
JSON metafile is created with suffix '.json'.
.I txtfile
Filename of text file.
.TP
.I clmfile
Filename of clm data file written.
.SH EXAMPLES
.TP
Create clm data from text file:
.B txt2clm
-nbands 365 tmp_1901-2003.txt tmp2003.clm
.PP
.SH EXIT STATUS
.B txt2clm
returns a zero exit status if the clm file is created.
Non zero is returned in case of failure.

.SH AUTHORS

For authors and contributors see AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
lpjml(1), asc2clm(1), cru2clm(1), grid2clm(1), txt2grid.(1), cft2clm(1), catclm(1), clm(5)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security

## [Unreleased]

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code reviewe: Stephen Wirth (wirth@pik-potsdam.de), Sibyll Schaphoff (sibylls@pik-potsdam.de)

### Added

- Utility `restart2yaml` added to convert the new restart files into human readable YAML or JSON files.
- Utility `json2restart` added to convert JSON files into restart files.
- alias `restart2json` added for `restart2yaml -json`.
- Output of time spent in reading and writing the restart file added.

### Changed

- Restart file format changed to include metadata. A binary JSON-like structure has been implemented. The order of the objects read can be different from the order objects have been written, but performance of reading can be degraded.
- Utility `lpjcat` updated to support the new restart file format.


## [5.9.25] - 2025-05-22

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Stephen Wirth (wirth@pik-potsdam.de), Fabian Stenzel (stenzel@pik-potsdam.de), Sebastian Ostberg(ostberg@pik-potsdam.de)

### Changed

- Axis names and missing values for NetCDF output are read from JSON file `netcdf.cjson` instead of being hard coded.
- Warnings in `addheader` replaced by terminating errors.
- Function `openinputfile()` returns an error if cell size differs in JSON metafile.
- Function `openmetafile()` returns an error if binary file is empty.
- Output format for cell size updated in `fprintheader.c`, `fprintoutputjson.c`, `fprintjson.c`, and `openinputfile.c`.

### Added

- New utility `splitclm` added to copy specific bands from a clm file.
- Environment variables `LPJNOPP` and `LPJPEDANTIC` added. If set to `true`, preprocessor is disabled and pedantic checking is enabled, respectively.
- Info about `--help` option added in case of error in command line arguments.
- Option `-config` added to `cdf2clm`, `bin2cdf`, `country2cdf` and `clm2cdf` in order to read axis names and missing values from a JSON file. File `netcdf.json` added as a template.
- Option `-json` added to `arr2clm`, `asc2clm`, `txt2grid`, `txt2clm`, `getcountry`, and `cdf2coord` in order to create additional JSON metafiles.
- Option `-inpath` added to `regridlpj` in order to set directory where data can be found.
- Options `-zero` and `-search` added to `regridlpj`.
- Option `-double` added to `txt2grid` in order to create grid files with coordinate values of type double.
- Option `-timestep` added to `addheader`, `txt2clm` and `bin2cdf`.
- Option `sum`, `tsum` and `tmean` to `mathclm` added.
- Check for correct setting of nstep and nbands added in `txt2clm`.

### Fixed

- Misspelled `-fsanitize=address` option corrected in `Makefile.mpich`.
- Missing `free(cmd)` added in `fprintincludes.c`.
- Parsing of options corrected in `fprintincludes.c`.
- Utility `bin2cdf` fixed for timestep>1.


## [5.9.24] - 2025-05-15

### Contributors

- author: David Hötten (davidho@pik-potsdam.de)
- code review: Sebastian Ostberg (ostberg@pik-potsdam.de), Maik Billing (billing@pik-potsdam.de)

### Added

- Merge request template ``.gitlab/merge_request_templates/default.md`` that includes several checklists.


## [5.9.23] - 2025-05-08

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Jens Heinke (heinke@pik-potsdam.de), Fabian Stenzel(stenzel@pik-potsdam.de)

### Changed

- Option `-check` of `configure.sh` enables now run-time checks of memory leaks and memory access out of bounds and undefined variables for gcc and icx compiler. Optimization is not disabled. Use options `-debug -check` to disable.
- Datatype `List` used in `fprintfiles.c`.
- All allocated memory is now freed in utilities.
- Prescribed landcover input file changed in `input.cjson` to a file with the same number of natural PFTs.
- All remaining unsafe calls to `sprintf()` replaced by calls to `snprinf()` or `getsprinf()`.

### Fixed

- Check for missing values for integer input corrected in `readclimate_netcdf.c`.
- Memory leak fixed in `filesexist.c`.
- Calculation of global area fixed in `printglobal.c`.
- Check for null attribute pointer added in `fprintjson.c`.
- Uninitialized grid name initialized in `mathclm.c`.
- Call to `snprintf()` replaced by `getsprintf()` in `reservoir2cdf.c`.
- Datatype for index and writing the header corrected in `lpjcat.c`.
- Functions for qsort(), bisect(), and leftmostzero() changed to avoid run-time errors with `-check` option.
- Pointer set to NULL  if `k_est`is NULL in `initmanage.c`.
- Missing argument added in opening aquifer file for IMAGE.
- Scaling of coordinates fixed in `joingrid.c`.
- Fraction pointer initialized to NULL in `freadresdata.c` to avoid SEGV in `lpjprint.c` if reservoir data is read from restart file.
- Index set correctly for outflow cells in `regriddrain.c`.
- Check for maximum discharge length corrected in `printdrain.c`.
- Number of years of clm file set to 1 in `grid2clm.c`.
- Function `fscanlandcovermap.c` corrected (issue https://github.com/PIK-LPJmL/LPJmL/issues/48).
- Access out of bounds for `soil->freezdepth` fixed in `soiltemp.c`.
- Format specifier changed in `cat2bsq.c` and `cutclm.c` to compile without warnings using the clang compiler (issue #372).


## [5.9.22] - 2025-03-14

### Contributors


src/utils/txt2clm.c
/**************************************************************************************/
/**                                                                                \n**/
/**                      t  x  t  2  c  l  m  .  c                                 \n**/
/**                                                                                \n**/
/**     Converts text files into LPJ climate data files                            \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/

#undef USE_MPI /* no MPI */

#include "lpj.h"

#define USAGE "\nUsage: txt2clm [-h] [-version v] [-cellindex] [-scale s] [-float] [-int] [-nbands n] [-nstep n] [-timestep t] [-cellsize size]\n               [-firstcell n] [-ncell n] [-firstyear f] [-header id] [-csv c] [-json] txtfile clmfile\n"
#define ERR_USAGE USAGE "\nTry \"txt2clm --help\" for more information.\n"

static int getfloat(FILE *file,char sep,float *value)
{
  String s;
  int c,len;
  char *endptr;
  len=0;
  while((c=fgetc(file))!=EOF)
  {
    if(len==STRING_LEN)
      return 0;
    if(c=='\n' || c==sep)
    {
      s[len]='\0';
      *value=(float)strtod(s,&endptr);
      if(*endptr=='\0')
        return 1;
      fprintf(stderr,"Invalid number '%s'.\n",s);
      return 0;
    }
    s[len++]=(char)c;
  }
  return 0;
} /* of 'getfloat' */

static int getint(FILE *file,char sep,int *value)
{
  String s;
  int c,len;
  char *endptr;
  len=0;
  while((c=fgetc(file))!=EOF)
  {
    if(len==STRING_LEN)
      return 0;
    if(c=='\n' || c==sep)
    {
      s[len]='\0';
      *value=strtol(s,&endptr,10);
      if(*endptr=='\0')
        return 1;
      fprintf(stderr,"Invalid number '%s'.\n",s);
      return 0;
    }
    s[len++]=(char)c;
  }
  return 0;
} /* of 'getint' */

int main(int argc,char **argv)
{
  FILE *file,*out;
  float multiplier;
  char *endptr;
  float value;
  short s;
  int data;
  Header header;
  int version;
  char *id;
  char *arglist,*out_json;
  int i,iarg,rc;
  Bool isjson;
  char sep;
  /* set default values */
  header.order=CELLYEAR;
  header.firstyear=1901;
  header.nbands=12;
  header.nstep=1;
  header.timestep=1;
  header.firstcell=0;
  header.ncell=67420;
  multiplier=1;
  header.datatype=LPJ_SHORT;
  header.cellsize_lon=header.cellsize_lat=0.5;
  id=LPJ_CLIMATE_HEADER;
  version=LPJ_CLIMATE_VERSION;
  isjson=FALSE;
  sep='\0';
  /* parse command line options */
  for(iarg=1;iarg<argc;iarg++)
    if(argv[iarg][0]=='-')
    {
      if(!strcmp(argv[iarg],"-h") || !strcmp(argv[iarg],"--help"))
      {
        printf("   txt2clm (" __DATE__ ") Help\n"
               "   ==========================\n\n"
               "Convert text files to clm data files for LPJmL version %s\n",getversion());
        printf(USAGE
               "\nArguments:\n"
               "-h,--help    print this help text\n"
               "-version     set version, default is %d\n"
               "-cellindex   set order to cell index\n"
               "-csv c       file is in CSV format with separator c\n"
               "-float       write data as float, default is short\n"
               "-int         write data as int, default is short\n"
               "-nbands n    number of bands, default is %d\n"
               "-nstep n     number of steps, default is %d\n"
               "-timestep n  number of years between time step, defauls is %d\n"
               "-firstcell n index of first cell\n"
               "-ncell n     number of cells, default is %d\n"
               "-firstyear f first year, default is %d\n"
               "-scale s     scale data by a factor of s\n"
               "-cellsize s  cell size, default is %g\n"
               "-header id   clm header string, default is '%s'\n"
               "-json        JSON metafile is created with suffix '.json'\n"
               "txtfile      filename of text file\n"
               "clmfile      filename of clm data file\n\n"
               "(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file\n",
               version,header.nbands,header.nstep,header.timestep,header.ncell,header.firstyear,header.cellsize_lon,id);
        return EXIT_SUCCESS;
      }
      else if(!strcmp(argv[iarg],"-float"))
        header.datatype=LPJ_FLOAT;
      else if(!strcmp(argv[iarg],"-int"))
        header.datatype=LPJ_INT;
      else if(!strcmp(argv[iarg],"-cellindex"))
        header.order=CELLINDEX;
      else if(!strcmp(argv[iarg],"-json"))
        isjson=TRUE;
      else if(!strcmp(argv[iarg],"-scale"))
      {
        if(iarg==argc-1)
        {
          fputs("Argument missing after '-scale' option.\n"
                ERR_USAGE,stderr);
          return EXIT_FAILURE;
        }
        multiplier=(float)strtod(argv[++iarg],&endptr);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid value '%s' for option '-scale'.\n",
                  argv[iarg]);
          return EXIT_FAILURE;
        }
        if(multiplier<=0)
        {
          fprintf(stderr,"Scale=%g must be greater than zero.\n",multiplier);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[iarg],"-cellsize"))
      {
        if(iarg==argc-1)
        {
          fputs("Argument missing after '-cellsize' option.\n"
                ERR_USAGE,stderr);
          return EXIT_FAILURE;
        }
        header.cellsize_lon=header.cellsize_lat=(float)strtod(argv[++iarg],&endptr);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid value '%s' for option '-cellsize'.\n",
                  argv[iarg]);
          return EXIT_FAILURE;
        }
        if(header.cellsize_lon<=0)
        {
          fprintf(stderr,"Cell size=%g must be greater than zero.\n",header.cellsize_lon);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[iarg],"-nbands"))
      {
        if(iarg==argc-1)
        {
          fputs("Argument missing after '-nbands' option.\n"
                ERR_USAGE,stderr);
          return EXIT_FAILURE;
        }
        header.nbands=strtol(argv[++iarg],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid value '%s' for option '-nbands'.\n",
                  argv[iarg]);
          return EXIT_FAILURE;
        }
        if(header.nbands<1)
        {
          fprintf(stderr,"Number of bands=%d must be greater than zero.\n",header.nbands);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[iarg],"-csv"))
      {
        if(iarg==argc-1)
        {
          fputs("Argument missing after '-csv' option.\n"
                ERR_USAGE,stderr);
          return EXIT_FAILURE;
        }
        sep=argv[++iarg][0];
      }
      else if(!strcmp(argv[iarg],"-nstep"))
      {
        if(iarg==argc-1)
        {
          fputs("Argument missing after '-nstep' option.\n"
                ERR_USAGE,stderr);
          return EXIT_FAILURE;
        }
        header.nstep=strtol(argv[++iarg],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid value '%s' for option '-nstep'.\n",
                  argv[iarg]);
          return EXIT_FAILURE;
        }
        if(header.nstep!=1 && header.nstep!=NMONTH && header.nstep!=NDAYYEAR)
        {
           fprintf(stderr,"Error: Number of steps=%d must be 1, 12, or 365.\n",header.nstep);
           return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[iarg],"-timestep"))
      {
        if(iarg==argc-1)
        {
          fputs("Argument missing after '-timestep' option.\n"
                ERR_USAGE,stderr);
          return EXIT_FAILURE;
        }
        header.timestep=strtol(argv[++iarg],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid value '%s' for option '-timestep'.\n",
                  argv[iarg]);
          return EXIT_FAILURE;
        }
        if(header.timestep<1)
        {
          fprintf(stderr,"Number of time steps=%d must be greater than zero.\n",header.timestep);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[iarg],"-version"))
      {
        if(iarg==argc-1)
        {
          fputs("Argument missing after '-version' option.\n"
                ERR_USAGE,stderr);
          return EXIT_FAILURE;
        }
        version=strtol(argv[++iarg],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid value '%s' for option '-version'.\n",
                  argv[iarg]);
          return EXIT_FAILURE;
        }
        if(version<1)
        {
          fprintf(stderr,"Version=%d must be greater than zero.\n",version);
          return EXIT_FAILURE;
        }
        if(version>CLM_MAX_VERSION)
        {
          fprintf(stderr,"Version %d greater than maximum version %d supported.\n",
                  version,CLM_MAX_VERSION);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[iarg],"-firstcell"))
      {
        if(iarg==argc-1)
        {
          fputs("Argument missing after '-firstcell' option.\n"
                ERR_USAGE,stderr);
          return EXIT_FAILURE;
        }
        header.firstcell=strtol(argv[++iarg],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid value '%s' for option '-firstcell'.\n",
                  argv[iarg]);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[iarg],"-ncell"))
      {
        if(iarg==argc-1)
        {
          fputs("Argument missing after '-ncell' option.\n"
                ERR_USAGE,stderr);
          return EXIT_FAILURE;
        }
        header.ncell=strtol(argv[++iarg],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid value '%s' for option '-ncell'.\n",
                  argv[iarg]);
          return EXIT_FAILURE;
        }
        if(header.ncell<1)
        {
          fprintf(stderr,"Number of cells=%d must be greater than zero.\n",header.ncell);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[iarg],"-firstyear"))
      {
        if(iarg==argc-1)
        {
          fputs("Argument missing after '-firstyear' option.\n"
                ERR_USAGE,stderr);
          return EXIT_FAILURE;
        }
        header.firstyear=strtol(argv[++iarg],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid value '%s' for option '-firstyear'.\n",
                  argv[iarg]);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[iarg],"-header"))
      {
        if(iarg==argc-1)
        {
          fputs("Argument missing after '-header' option.\n"
                ERR_USAGE,stderr);
          return EXIT_FAILURE;
        }
        id=argv[++iarg];
      }
      else
      {
        fprintf(stderr,"Error: Invalid option '%s'.\n",argv[iarg]);
        fprintf(stderr,ERR_USAGE);
        return EXIT_FAILURE;
      }
    }
    else
      break;
  if(argc<iarg+2)
  {
    fputs("Filename(s) missing.\n"
          ERR_USAGE,stderr);
    return EXIT_FAILURE;
  }
  if(header.timestep>1 && header.nstep>1)
  {
    fprintf(stderr,"Error: Number of time steps=%d within a year must be 1 if years between time steps=%d is greater than 1.\n",
            header.nstep,header.timestep);
    return EXIT_FAILURE;
  }
  if(version<4)
  {
    if(header.nstep>1 && header.nbands==1)
    {
      fprintf(stderr,"Warning: Version of clm file is %d and nstep>1, nbands set to nstep=%d\n",
              version,header.nstep);
      header.nbands=header.nstep;
    }
    else if(header.nstep>1 && header.nbands>1)
    {
      fprintf(stderr,"Error: Number of steps=%d and bands=%d must not be both >1 for verion %d clm files.\n",
              header.nstep,header.nbands,version);
      return EXIT_FAILURE;
    }
    if(header.timestep>1)
      fprintf(stderr,"Warning: Version of clm file is %d and timestep>1, timestep is not written to clm file.\n",version);
  }
  file=fopen(argv[iarg],"r");
  if(file==NULL)
  {
    fprintf(stderr,"Error opening '%s': %s\n",argv[iarg],strerror(errno));
    return EXIT_FAILURE;
  }
  header.nyear=0;
  header.scalar=1/multiplier;
  out=fopen(argv[iarg+1],"wb");
  if(out==NULL)
  {
    fprintf(stderr,"Error creating '%s': %s\n",argv[iarg+1],strerror(errno));
    return EXIT_FAILURE;
  }
  fwriteheader(out,&header,id,version);
  if(header.order==CELLINDEX)
  {
    for(i=0;i<header.ncell;i++)
    {
      if(sep)
        rc=getint(file,sep,&data);
      else
        rc=fscanf(file,"%d",&data);
      if(rc!=1)
      {
        fprintf(stderr,"Unexpected end of file in '%s' reading cell index.\n",argv[iarg]);
        return EXIT_FAILURE;
      }
      if(fwrite(&data,sizeof(int),1,out)!=1)
      {
        fprintf(stderr,"Error writing data in '%s': %s.\n",argv[iarg+1],strerror(errno));
        return EXIT_FAILURE;
      }
    }
  }
  do
  {
    if(header.datatype==LPJ_INT)
    {
      for(i=0;i<header.ncell*header.nbands*header.nstep;i++)
      {
        if(sep)
          rc=getint(file,sep,&data);
        else
          rc=fscanf(file,"%d",&data);
        if(rc!=1)
          break;
        if(fwrite(&data,sizeof(data),1,out)!=1)
        {
          fprintf(stderr,"Error writing data in '%s': %s.\n",argv[iarg+1],strerror(errno));
          return EXIT_FAILURE;
        }
      }
    }
    else
      for(i=0;i<header.ncell*header.nbands*header.nstep;i++)
      {
        if(sep)
          rc=getfloat(file,sep,&value);
        else
          rc=fscanf(file,"%g",&value);
        if(rc!=1)
          break;
        if(header.datatype==LPJ_SHORT)
        {
          if(value*multiplier<SHRT_MIN || value*multiplier>SHRT_MAX)
             fprintf(stderr,"WARNING: Data overflow %g in year %d at cell %d and band %d\n",
                     value*multiplier,header.nyear,i/header.nbands/header.nstep+1,i % (header.nbands*header.nstep)+1);
          s=(short)(value*multiplier);
          if(fwrite(&s,sizeof(short),1,out)!=1)
          {
            fprintf(stderr,"Error writing data in '%s': %s.\n",argv[iarg+1],strerror(errno));
            return EXIT_FAILURE;
          }
        }
        else
        {
          if(fwrite(&value,sizeof(float),1,out)!=1)
          {
            fprintf(stderr,"Error writing data in '%s': %s.\n",argv[iarg+1],strerror(errno));
            return EXIT_FAILURE;
          }
        }
    }
    if(i==header.ncell*header.nbands*header.nstep)
      header.nyear++;
    else if(i!=0)
      fprintf(stderr,"Cannot read data in year %d at cell %d and band %d.\n",
              header.nyear+1,i / header.nbands/header.nstep+1,i % (header.nbands*header.nstep)+1);
    else if(!feof(file))
      fprintf(stderr,"End of file not reached in '%s'.\n",argv[iarg]);
  } while(rc==1);
  fclose(file);
  rewind(out);
  fwriteheader(out,&header,id,version);
  fclose(out);
  if(isjson)
  {
    out_json=malloc(strlen(argv[iarg+1])+strlen(JSON_SUFFIX)+1);
    if(out_json==NULL)
    {
      printallocerr("filename");
      return EXIT_FAILURE;
    }
    strcat(strcpy(out_json,argv[iarg+1]),JSON_SUFFIX);
    arglist=catstrvec(argv,argc);
    out=fopen(out_json,"w");
    if(out==NULL)
    {
      printfcreateerr(out_json);
      return EXIT_FAILURE;
    }
    fprintjson(out,argv[iarg+1],NULL,"txt2clm",NULL,arglist,&header,NULL,NULL,NULL,0,NULL,NULL,NULL,NULL,NULL,LPJ_SHORT,CLM,id,FALSE,version);
    fclose(out);
  }
  return EXIT_SUCCESS;
} /* of 'main' */
